README.md
# t3ext-jk_poll


ChangeLog
2015-03-30 Hielke Jager [WIND Internet] <info@windinternet.nl>
	* Version 0.9.99
	* Changed configuration. Added option to overrule the default storage folder in the plugin.

tca.php
<?php
if (!defined ("TYPO3_MODE")) 	die ("Access denied.");


$TCA["tx_jkpoll_poll"]["ctrl"]["dividers2tabs"] = 1;
$TCA["tx_jkpoll_poll"] = Array (
	"ctrl" => $TCA["tx_jkpoll_poll"]["ctrl"],
	"interface" => Array (
		"showRecordFieldList" => "hidden,starttime,endtime,fe_group,title,image,question,votes,answers,colors,crdate"
	),
	"feInterface" => $TCA["tx_jkpoll_poll"]["feInterface"],
	"columns" => Array (
		"crdate" => Array (		
			"exclude" => 1,
			"l10n_mode" => "mergeIfNotBlank",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.crdate",
			"config" => Array (
				"type" => "input",
				"size" => "15",
				"eval" => "date",
			)
		),
		"hidden" => Array (		
			"exclude" => 1,
			"l10n_mode" => "mergeIfNotBlank",
			"label" => "LLL:EXT:lang/locallang_general.php:LGL.hidden",
			"config" => Array (
				"type" => "check",
				"default" => "0"
			)
		),
		"starttime" => Array (		
			"exclude" => 1,
			"l10n_mode" => "mergeIfNotBlank",
			"label" => "LLL:EXT:lang/locallang_general.php:LGL.starttime",
			"config" => Array (
				"type" => "input",
				"size" => "8",
				"max" => "20",
				"eval" => "date",
				"default" => "0",
				"checkbox" => "0"
			)
		),
		"endtime" => Array (		
			"exclude" => 1,
			"l10n_mode" => "mergeIfNotBlank",
			"label" => "LLL:EXT:lang/locallang_general.php:LGL.endtime",
			"config" => Array (
				"type" => "input",
				"size" => "8",
				"max" => "20",
				"eval" => "date",
				"checkbox" => "0",
				"default" => "0",
				"range" => Array (
					"upper" => mktime(0,0,0,12,31,2020),
					"lower" => mktime(0,0,0,date("m")-1,date("d"),date("Y"))
				)
			)
		),
		"fe_group" => Array (		
			"exclude" => 1,
			"l10n_mode" => "mergeIfNotBlank",
			"label" => "LLL:EXT:lang/locallang_general.php:LGL.fe_group",
			"config" => Array (
				"type" => "select",
				"items" => Array (
					Array("", 0),
					Array("LLL:EXT:lang/locallang_general.php:LGL.hide_at_login", -1),
					Array("LLL:EXT:lang/locallang_general.php:LGL.any_login", -2),
					Array("LLL:EXT:lang/locallang_general.php:LGL.usergroups", "--div--")
				),
				"foreign_table" => "fe_groups"
			)
		),
		"title" => Array (		
			"exclude" => 1,
			"l10n_mode" => "prefixLangTitle",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.title",		
			"config" => Array (
				"type" => "input",	
				"size" => "30",	
				"eval" => "required",
			)
		),
		"image" => Array (		
			"exclude" => 1,
			"l10n_mode" => "exclude",		
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.image",		
			"config" => Array (
				"type" => "group",
				"internal_type" => "file",
				"allowed" => $GLOBALS["TYPO3_CONF_VARS"]["GFX"]["imagefile_ext"],	
				"max_size" => 500,	
				"uploadfolder" => "uploads/tx_jkpoll",
				"show_thumbs" => 1,	
				"size" => 1,	
				"minitems" => 0,
				"maxitems" => 1,
			)
		),
		"question" => Array (		
			"exclude" => 1,
			"l10n_mode" => "prefixLangTitle",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.question",		
			"config" => Array (
				"type" => "text",
				"cols" => "30",
				"rows" => "5",
				"wizards" => Array(
					"_PADDING" => 2,
					"RTE" => Array(
						"notNewRecords" => 1,
						"RTEonly" => 1,
						"type" => "script",
						"title" => "Full screen Rich Text Editing|Formatteret redigering i hele vinduet",
						"icon" => "wizard_rte2.gif",
						"script" => "wizard_rte.php",
					),
				),
			)
		),
		"votes" => Array (		
			"exclude" => 1,
			"l10n_mode" => "exclude",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.votes",
			"config" => Array (
				"type" => "none",
				"cols" => "30",	
				"rows" => "5",
			)
		),
		"answers" => Array (		
			"exclude" => 1,
			"l10n_mode" => "prefixLangTitle",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.answers",		
			"config" => Array (
				"type" => "text",
				"cols" => "30",	
				"rows" => "5",
			)
		),
		"colors" => Array (		
			"exclude" => 1,
			"l10n_mode" => "mergeIfNotBlank",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.colors",		
			"config" => Array (
				"type" => "text",
				"cols" => "30",	
				"rows" => "5",
			)
		),
		"votes_count" => Array (		
			"exclude" => 1,
			"l10n_mode" => "exclude",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.votes_count",
			"config" => Array (
				"type" => "none",
				"cols" => "30",	
			)
		),
		"valid_till" => Array (        
			"exclude" => 1,
			"l10n_mode" => "mergeIfNotBlank",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.valid_till",        
			"config" => Array (
				"type" => "input",
				"size" => "8",
				"max" => "20",
				"eval" => "date",
				"checkbox" => "0",
				"default" => "0"
			),
		),        
		"title_tag" => Array (        
			"exclude" => 1,
			"l10n_mode" => "prefixLangTitle",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.title_tag",        
			"config" => Array (
				"type" => "text",
				"cols" => "30",
				"rows" => "2",
			)
		),
		"alternative_tag" => Array (        
			"exclude" => 1,
			"l10n_mode" => "prefixLangTitle",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.alternative_tag",        
			"config" => Array (
				"type" => "text",
				"cols" => "30",
				"rows" => "2",
			)
		),
		"width" => Array (        
			"exclude" => 1,
			"l10n_mode" => "exclude",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.width",        
			"config" => Array (
				"type" => "input",
				"size" => "4",
				"max" => "4",
				"eval" => "int",
				"checkbox" => "0",
				"range" => Array (
					"upper" => "1000",
					"lower" => "10"
				),
				"default" => 0
			)
		),
		"height" => Array (        
			"exclude" => 1,
			"l10n_mode" => "exclude",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.height",        
			"config" => Array (
				"type" => "input",
				"size" => "4",
				"max" => "4",
				"eval" => "int",
				"checkbox" => "0",
				"range" => Array (
					"upper" => "1000",
					"lower" => "10"
				),
				"default" => 0
			)
		),
		"link" => Array (        
			"exclude" => 1,
			"l10n_mode" => "mergeIfNotBlank",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.link",        
			"config" => Array (
				"type" => "input",
				"size" => "15",
				"max" => "255",
				"checkbox" => "",
				"eval" => "trim",
				"wizards" => Array(
				"_PADDING" => 2,
				"link" => Array(
						"type" => "popup",
						"title" => "Link",
						"icon" => "link_popup.gif",
						"script" => "browse_links.php?mode=wizard",
						"JSopenParams" => "height=300,width=500,status=0,menubar=0,scrollbars=1"
					)
				)
			)
		),
		"clickenlarge" => Array (        
			"exclude" => 1,
			"l10n_mode" => "exclude",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.clickenlarge",        
			"config" => Array (
				"type" => "check",
			)
		),
		"answers_image" => Array (		
			"exclude" => 1,
			"l10n_mode" => "exclude",		
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.answers_image",		
			"config" => Array (
				"type" => "group",
				"internal_type" => "file",
				"allowed" => $GLOBALS["TYPO3_CONF_VARS"]["GFX"]["imagefile_ext"],	
				"max_size" => 500,	
				"uploadfolder" => "uploads/tx_jkpoll",
				"show_thumbs" => 1,	
				"size" => 3,	
				"minitems" => 0,
				"maxitems" => 20,
			)
		),
		"answers_description" => Array (		
			"exclude" => 1,
			"l10n_mode" => "prefixLangTitle",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.answers_description",		
			"config" => Array (
				"type" => "text",
				"cols" => "30",	
				"rows" => "5",
			)
		),
		"explanation" => Array (		
			"exclude" => 1,
			"l10n_mode" => "prefixLangTitle",
			"label" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.explanation",		
			"config" => Array (
				"type" => "text",
				"cols" => "30",
				"rows" => "3",
				"wizards" => Array(
					"_PADDING" => 2,
					"RTE" => Array(
						"notNewRecords" => 1,
						"RTEonly" => 1,
						"type" => "script",
						"title" => "Full screen Rich Text Editing|Formatteret redigering i hele vinduet",
						"icon" => "wizard_rte2.gif",
						"script" => "wizard_rte.php",
					),
				),
			)
		),
		'sys_language_uid' => Array(
			'exclude' => 1,
			'label' => 'LLL:EXT:lang/locallang_general.php:LGL.language',
			'config' => Array(
				'type' => 'select',
				'foreign_table' => 'sys_language',
				'foreign_table_where' => 'ORDER BY sys_language.title',
				'items' => Array(
					Array('LLL:EXT:lang/locallang_general.php:LGL.allLanguages',-1),
					Array('LLL:EXT:lang/locallang_general.php:LGL.default_value',0)
				)
			)
		),
		'l18n_parent' => Array(
			'displayCond' => 'FIELD:sys_language_uid:>:0',
			'exclude' => 1,
			'label' => 'LLL:EXT:lang/locallang_general.php:LGL.l18n_parent',
			'config' => Array(
				'type' => 'select',
				'items' => Array(
					Array('', 0),
				),
				'foreign_table' => 'tx_jkpoll_poll',
				'foreign_table_where' => 'AND tx_jkpoll_poll.uid=###REC_FIELD_l18n_parent### AND tx_jkpoll_poll.sys_language_uid IN (-1,0)',
				'wizards' => Array(
					'_PADDING' => 2,
					'_VERTICAL' => 1,
					'edit' => Array(
						'type' => 'popup',
						'title' => 'edit default language version of this record ',
						'script' => 'wizard_edit.php',
						'popup_onlyOpenIfSelected' => 1,
						'icon' => 'edit2.gif',
						'JSopenParams' => 'height=600,width=700,status=0,menubar=0,scrollbars=1,resizable=1',
					)
				)
			)
		),
		'l18n_diffsource' => Array(
			'config' => Array(
				'type'=>'passthrough'
			)
		),
	),
	"types" => Array (
		"0" => Array("showitem" => 
					'crdate, title, question;;4;richtext[paste|bold|italic|underline|formatblock|class|left|center|right|orderedlist|unorderedlist|outdent|indent|link|image]:rte_transform[mode=ts], answers;;2, colors, explanation;;;richtext[paste|bold|italic|underline|formatblock|class|left|center|right|orderedlist|unorderedlist|outdent|indent|link|image]:rte_transform[mode=ts];1-1-1,
					--div--;LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.tabs.image, image;;3, title_tag, alternative_tag, width, height,
					--div--;LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.tabs.answers, answers_image, answers_description,
					--div--;LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll.tabs.access, hidden;;1, valid_till,
					')
	
	
	/*			"hidden;;1;;1-1-1, title;;;;1-1-1,question;;;richtext[paste|bold|italic|underline|formatblock|class|left|center|right|orderedlist|unorderedlist|outdent|indent|link|image]:rte_transform[mode=ts],votes_count;;2,votes,answers,colors,answers_image,answers_description,explanation;;;richtext[paste|bold|italic|underline|formatblock|class|left|center|right|orderedlist|unorderedlist|outdent|indent|link|image]:rte_transform[mode=ts], image;;;;2-2-2,title_tag,alternative_tag,width,height,link,clickenlarge,crdate")
	
	'0' => Array('showitem' =>
            'hidden, type;;;;1-1-1,title;;;;2-2-2,short,bodytext;;2;richtext:rte_transform[flag=rte_enabled|mode=ts];4-4-4,
            --div--;LLL:EXT:tt_news/locallang_tca.xml:tt_news.tabs.special, datetime;;;;2-2-2,archivedate,author;;3;; ;;;;2-2-2,
                keywords;;;;2-2-2,sys_language_uid;;1;;3-3-3,
            --div--;LLL:EXT:tt_news/locallang_tca.xml:tt_news.tabs.media, image;;;;1-1-1,imagecaption;;5;;,links;;;;2-2-2,news_files;;;;4-4-4,
            --div--;LLL:EXT:tt_news/locallang_tca.xml:tt_news.tabs.catAndRels, category;;;;3-3-3,related;;;;3-3-3,
            --div--;LLL:EXT:tt_news/locallang_tca.xml:tt_news.tabs.access, starttime,endtime,fe_group,editlock,
            --div--;LLL:EXT:tt_news/locallang_tca.xml:tt_news.tabs.extended,
            '),*/
	
		
	),
	"palettes" => Array (
		"1" => Array("showitem" => "starttime, endtime, fe_group"),
		"2" => Array("showitem" => "votes_count, votes"),
	 	"3" => Array("showitem" => "link, clickenlarge"),
//		"4" => Array("showitem" => "explanation;;;richtext[paste|bold|italic|underline|formatblock|class|left|center|right|orderedlist|unorderedlist|outdent|indent|link|image]:rte_transform[mode=ts]")
	)
);
?>


ext_tables.php
<?php
if (!defined ("TYPO3_MODE")) 	die ("Access denied.");

t3lib_extMgm::allowTableOnStandardPages("tx_jkpoll_poll");

$TCA["tx_jkpoll_poll"] = Array (
	"ctrl" => Array (
		"title" => "LLL:EXT:jk_poll/locallang_db.xml:tx_jkpoll_poll",		
		"label" => "title",	
		"default_sortby" => "ORDER BY crdate DESC",
		"tstamp" => "tstamp",
		"crdate" => "crdate",
		"cruser_id" => "cruser_id",
		"delete" => "deleted",	
		"enablecolumns" => Array (		
			"disabled" => "hidden",	
			"starttime" => "starttime",	
			"endtime" => "endtime",	
			"fe_group" => "fe_group",
		),
		"dynamicConfigFile" => t3lib_extMgm::extPath($_EXTKEY)."tca.php",
		"iconfile" => t3lib_extMgm::extRelPath($_EXTKEY)."icon_tx_jkpoll_poll.gif",
		
		"prependAtCopy" => "LLL:EXT:lang/locallang_general.php:LGL.prependAtCopy",
		"copyAfterDuplFields" => "sys_language_uid",
		"useColumnsForDefaultValues" => "sys_language_uid",
		"transOrigPointerField" => "l18n_parent",
		"transOrigDiffSourceField" => "l18n_diffsource",
		"languageField" => "sys_language_uid",

	),
	"feInterface" => Array (
		"fe_admin_fieldList" => "hidden, starttime, endtime, fe_group, title, image, question, votes, answers, colors, valid_till, answers_image, answers_description, explanation",
	)
);


t3lib_div::loadTCA("tt_content");
$TCA["tt_content"]["types"]["list"]["subtypes_excludelist"][$_EXTKEY."_pi1"]="layout,select_key";
$TCA['tt_content']['types']['list']['subtypes_addlist'][$_EXTKEY.'_pi1']='pi_flexform';

t3lib_extMgm::addPlugin(Array("LLL:EXT:jk_poll/locallang_db.xml:tt_content.list_type_pi1", $_EXTKEY."_pi1"),"list_type");
t3lib_extMgm::addStaticFile($_EXTKEY,"pi1/static/","Poll");
t3lib_extMgm::addPiFlexFormValue($_EXTKEY.'_pi1', 'FILE:EXT:jk_poll/flexform_ds.xml');

if (TYPO3_MODE=="BE")	$TBE_MODULES_EXT["xMOD_db_new_content_el"]["addElClasses"]["tx_jkpoll_pi1_wizicon"] = t3lib_extMgm::extPath($_EXTKEY)."pi1/class.tx_jkpoll_pi1_wizicon.php";
?>


class.ext_update.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2004 Johannes Krausmueller (johannes@krausmueller.de)
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/
/**
 * Plugin 'Poll' for the 'jk_poll' extension.
 *
 * @author	Johannes Krausmueller <johannes@krausmueller.de>
*/

class ext_update  {

    /**
     * Main function, returning the HTML content of the module
     * 
     * @return    string        HTML
     */
    function main()    {
    	//select all polls
    	$res_polls = $GLOBALS['TYPO3_DB']->exec_SELECTquery('*', 'tx_jkpoll_poll','');
		while ($row_polls = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($res_polls)) {	
			//check if update is necessary
			$pos = strpos($row_polls['answers'], '|');
			if ($pos > 0) {
				// get every answer
				$answers = explode("\n", $row_polls['answers']);
				$answers_new = "";
				$votes_new = "";
				$colors_new = "";
				foreach ($answers as $answer) {				
					list ($votes, $text, $color) = explode('|', $answer);				
					$answers_new .=  $text."\n";
					$votes_new .= $votes."\n";
					$colors_new .= $color."\n";
				} 
				$answers_new =  trim($answers_new);
				$votes_new = trim($votes_new);
				$colors_new = trim($colors_new);
				
				//write back to database			
				$updateFields = array(
					'answers' => $answers_new,
					'votes' => $votes_new,
					'colors' => $colors_new
				);
				$GLOBALS['TYPO3_DB']->exec_UPDATEquery('tx_jkpoll_poll', 'uid='.$row_polls['uid'], $updateFields);
			}
		}        
		$content.= '<br />Update complete!';
        return $content;    
    }
    
    /**
     * Checks if update is needed
     * 
     * @return    boolean        
     */
    function access()    {

		//check if update is necessary
       	$res_polls = $GLOBALS['TYPO3_DB']->exec_SELECTquery('*', 'tx_jkpoll_poll','');
		while ($row_polls = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($res_polls)) {	
			$pos = strpos($row_polls['answers'], '|');
			if ($pos > 0) {
				$update = true;
			}			
    	}
    	if ($update == true) {
    		return $update;
		}
    }        
}
?>

pi1/class.tx_jkpoll_pi1_wizicon.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2004 Johannes Krausmueller (johannes@krausmueller.de)
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/
/**
 * Class that adds the wizard icon.
 *
 * @author	Johannes Krausmueller <johannes@krausmueller.de>
 */



class tx_jkpoll_pi1_wizicon {
	function proc($wizardItems)	{
		global $LANG;

		$LL = $this->includeLocalLang();

		$wizardItems["plugins_tx_jkpoll_pi1"] = array(
			"icon"=>t3lib_extMgm::extRelPath("jk_poll")."pi1/ce_wiz.gif",
			"title"=>$LANG->getLLL("pi1_title",$LL),
			"description"=>$LANG->getLLL("pi1_plus_wiz_description",$LL),
			"params"=>"&defVals[tt_content][CType]=list&defVals[tt_content][list_type]=jk_poll_pi1"
		);

		return $wizardItems;
	}
	function includeLocalLang()    {
        $llFile = t3lib_extMgm::extPath('jk_poll').'locallang_db.xml';
        //if (t3lib_div::int_from_ver(TYPO3_version) < 4007000) {
	if (version_compare(TYPO3_branch,'4.7','<')) {
        	$LOCAL_LANG = t3lib_div::readLLXMLfile($llFile, $GLOBALS['LANG']->lang);
        } else {
        	$llxmlParser = t3lib_div::makeInstance('t3lib_l10n_parser_Llxml');
		$LOCAL_LANG = $llxmlParser->getParsedData($llFile, $GLOBALS['LANG']->lang);
        }
        return $LOCAL_LANG;
    }
}



if (defined("TYPO3_MODE") && $TYPO3_CONF_VARS[TYPO3_MODE]["XCLASS"]["ext/jk_poll/pi1/class.tx_jkpoll_pi1_wizicon.php"])	{
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]["XCLASS"]["ext/jk_poll/pi1/class.tx_jkpoll_pi1_wizicon.php"]);
}

?>


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "jk_poll".
 *
 * Auto generated 30-03-2015 13:50
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array (
	'title' => 'Poll',
	'description' => 'A poll based on the extension quickpoll. A template-file can be used to define the output in the frontend. It is also possible to create a horiontal or vertical display of the percentage of users voted for an answer.',
	'category' => 'plugin',
	'version' => '0.9.99',
	'state' => 'beta',
	'uploadfolder' => false,
	'createDirs' => '',
	'clearcacheonload' => true,
	'author' => 'Johannes Krausmueller',
	'author_email' => 'johannes@krausmueller.de',
	'author_company' => '',
	'constraints' =>
	array (
		'depends' =>
		array (
			'typo3' => '4.5.0-6.2.99',
		),
		'conflicts' =>
		array (
		),
		'suggests' =>
		array (
		),
	),
);